proxy/coco/src/peer/run_state.rs

```diff
@@ -0,0 +1,361 @@
+//! State machine to manage the current mode of operation during peer lifecycle.
+
+use std::{collections::HashSet, time::Instant};
+
+use librad::{
+    net::{peer::Gossip, protocol::ProtocolEvent},
+    peer::PeerId,
+};
+
+use crate::peer::announcement;
+
+/// Number of peers a full sync is attempting with up on startup.
+/// TODO(xla): Revise number.
+/// TODO(xla): Make configurable as part of peer configuration.
+const SYNC_MAX_PEERS: usize = 5;
+
+/// Instructions to issue side-effectful operations which are the results from state transitions.
+#[derive(Debug, PartialEq)]
+pub enum Command {
+    /// Start the announcement subroutine.
+    Announce,
+    /// Initiate a full sync with `PeerId`.
+    SyncPeer(PeerId),
+    /// Start sync timeout.
+    StartSyncTimeout,
+}
+
+/// Significant events that occur during [`Peer`] lifetime.
+#[allow(clippy::large_enum_variant)]
+#[derive(Clone, Debug)]
+pub enum Event {
+    /// Announcement subroutine lifecycle events.
+    Announce(AnnounceEvent),
+    /// Events from the underlying coco protocol.
+    Protocol(ProtocolEvent<Gossip>),
+    /// Lifecycle events during peer sync operations.
+    PeerSync(SyncEvent),
+    /// Scheduled timeouts which can occur.
+    Timeout(TimeoutEvent),
+}
+
+/// Announcement subroutine lifecycle events.
+#[derive(Clone, Debug)]
+pub enum AnnounceEvent {
+    /// Operation failed.
+    Failed,
+    /// Operation succeeded and emitted the enclosed list of updates.
+    Succeeded(announcement::Updates),
+    /// The ticker duration has elapsed.
+    Tick,
+}
+
+/// Lifecycle events during peer sync operations.
+#[derive(Clone, Debug)]
+pub enum SyncEvent {
+    /// A sync has been initiated for `PeerId`.
+    Started(PeerId),
+    /// A sync has failed for `PeerId`.
+    Failed(PeerId),
+    /// A sync has succeeded for `PeerId`.
+    Succeeded(PeerId),
+}
+
+/// Scheduled timeouts which can occur.
+#[derive(Clone, Debug)]
+pub enum TimeoutEvent {
+    /// Grace period is over signaling that we should go offline, no matter how many syncs have
+    /// succeeded.
+    SyncPeriod,
+}
+
+/// Internal state representation the current mode of operation for the local peer.
```

Comment:
It's representing current operational status of the local peer. E.g. offline when no peers are connected or syncing when it came online and catches up with the network. What could be rephrased here to make that more clear?

Reply:
Well, the sentence seems off I think because of this portion "representation the current mode", I'm guessing a typo or a missed word.

But also "mode of operation" seems a bit vague. Like the variants describe states of syncing and connectivity which doesn't seem like a mode of operation to me.

Don't wanna bikeshed on this so I'm just gonna suggest to fix the sentence and you can do what you'd like with my above misunderstanding :smile: